avldb/src/main/scala/scorex/db/LDBVersionedStore.scala

```diff
@@ -0,0 +1,324 @@
+    val options = new WriteOptions()
+    options.sync(true) // Is it enough to provide fault tolerance???
+    options
+  }
+
+  def get(key: K): Option[V] = {
+    lock.readLock().lock()
+    try {
+      val b = db.get(key.data)
+      if (b == null) None else Some(ByteArrayWrapper(b))
+    } finally {
+      lock.readLock().unlock()
+    }
+  }
+
+  def getAll(consumer: (K, V) => Unit): Unit = {
+    lock.readLock().lock()
+    val iterator = db.iterator()
+    try {
+      iterator.seekToFirst()
+      while (iterator.hasNext) {
+        val n = iterator.next()
+        consumer(ByteArrayWrapper(n.getKey), ByteArrayWrapper(n.getValue))
+      }
+    } finally {
+      iterator.close()
+      lock.readLock().unlock()
+    }
+  }
+
+  private def newLSN(): Array[Byte] = {
+    lsn += 1
+    encodeLSN(lsn)
+  }
+
+  // Invert word to provide descending key order.
+  // Java implementation of LevelDB org.iq80.leveldb doesn't support iteration in backward direction.
+  private def decodeLSN(lsn: Array[Byte]): LSN = {
+    ~ByteBuffer.wrap(lsn).getLong
+  }
+
+  private def encodeLSN(lsn: LSN): Array[Byte] = {
+    val buf = ByteBuffer.allocate(8)
+    buf.putLong(~lsn)
+    buf.array()
+  }
+
+  private def getLastLSN: LSN = {
+    val iterator = undo.iterator
+    try {
+      iterator.seekToFirst()
+      if (iterator.hasNext) {
+        decodeLSN(iterator.peekNext().getKey)
+      } else {
+        0
+      }
+    } finally {
+      iterator.close()
+    }
+  }
+
+  def lastVersionID: Option[VersionID] = {
+    lastVersion
+  }
+
+  def versionIDExists(versionID: VersionID): Boolean = {
+    lock.readLock().lock()
+    try {
+      versions.contains(versionID)
+    } finally {
+      lock.readLock().unlock()
+    }
+  }
+
+  private def getAllVersions: ArrayBuffer[VersionID] = {
+    val versions = ArrayBuffer.empty[VersionID]
+    var lastVersion: VersionID = null
+    var lastLsn: LSN = 0
+    // We iterate in LSN descending order
+    val iterator = undo.iterator()
+    iterator.seekToFirst()
+    while (iterator.hasNext) {
+      val entry = iterator.next
+      val currVersion = deserializeUndo(entry.getValue).versionID
+      lastLsn = decodeLSN(entry.getKey)
+      if (!currVersion.equals(lastVersion)) {
+        versionLsn += lastLsn + 1 // this is first LSN of successor version
+        versions += currVersion
+        lastVersion = currVersion
+      }
+    }
+    iterator.close()
+    // As far as org.iq80.leveldb doesn't support iteration in reverse order, we have to iterate in the order
+    // of decreasing LSNs and then revert version list. For each version we store first (smallest) LSN.
+    versionLsn += lastLsn // first LSN of oldest version
+    versionLsn = versionLsn.reverse // LSNs should be in ascent order
+    versionLsn.remove(versionLsn.size - 1) // remove last element which corresponds to next assigned LSN
+    versions.reverse
+  }
+
+  case class Undo(versionID: VersionID, key: Array[Byte], value : Array[Byte])
```

Comment:
Please describe what this class represents (its role)

Reply:
Done